rfcs/0010-target-registered-compiler-flow-customisation.md

```diff
@@ -0,0 +1,186 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+As a user, you can pick from additional hooks to bypass certain behaviours of the `Target`:
+* `RelayToTIR` - Customize the lowering flow to TIR
+* `TIRToRuntime` - Customize code generation into a runtime module from TIR
+* `RelayToRuntime` - Full compilation flow from Relay to a runtime module
+
+To illustrate where the hooks are placed, please refer to the following diagram:
+
+![Diagram showing the splitting point of RelayToRuntime, RelayToTIR and TIRToRuntime](./assets/000x/bypass.png)
+
+These can be registered on targets using `set_attr`:
+```
+TVM_REGISTER_TARGET_KIND("cmsisnn", kDLCPU)
+    .set_attr<FTVMRelayToTIR>("RelayToTIR", CMSISNNLowering)
+    .set_attr<FTVMTIRToRuntime>("TIRToRuntime", CMSISNNCodeGen);
+
+TVM_REGISTER_TARGET_KIND("ethos-n", kDLCPU)
+    .set_attr<FTVMRelayToRuntime>("RelayToRuntime", EthosNCodeGen)
+    .set_attr<FTVMConstantUpdater>("UpdateConstants", EthosNConstantUpdater);
+```
+
+## Relay -> TIR
+With this change, this path splits, depending on whether you wanted to generate a full `Module` or introduce some specific TIR nodes into the code generation flow; the addition of the `RelayToTIR` hook allows you to write trivial external TIR generators such as calling out to a third party library:
+```c++
+void CallExternalLibraryInTIR(const GlobalVar& new_global_var, const Function& func) {
+    tir::Buffer x_buffer = tir::decl_buffer({8}, DataType::Float(32), "x");
+    tir::Var x_var("x", DataType::Handle());
+
+    Map<String, ObjectRef> dict_attrs;
+    dict_attrs.Set("global_symbol", new_global_var->name_hint);
+    dict_attrs.Set("tir.noalias", Bool(true));
+
+    Map<tir::Var, tir::Buffer> buffer_map = {{x_var, x_buffer}};
+    tir::Stmt body =
+        tir::Evaluate(tvm::tir::Call(DataType::Int(8), tir::builtin::call_extern(), {x->data}));
+
+    tir::PrimFunc replacement_func = tir::PrimFunc({x_var}, body, VoidType(),
+                                                    buffer_map, DictAttrs(dict_attrs));
+    replacement_func = WithAttr(replacement_func, ::tvm::attr::kTarget, host_target_);
+    ir_module_->Add(new_global_var, replacement_func);
+}
+```
+This is then registered on a target:
+```c++
+TVM_REGISTER_TARGET_KIND("woofles", kDLCPU)
+    .set_attr<FTVMRelayToTIR>("RelayToTIR", relay::contrib::woofles::RelayToTIR());
+```
+The signature for this hook is as the same as any other `Pass`, which takes an `IRModule` with `Function`s and returns an `IRModule` with transformed `PrimFunc`s.
+
+## TIR -> Runtime
+Extending from the above, a second hook is introduced to do further transformations from TIR -> Runtime named `TIRToRuntime`, this bypasses the default `target.build.X` and instead uses the registered `TIRToRuntime` build:
+```c++
+runtime::Module BuildWooflesHost(IRModule mod, Target target) {
```

Comment:
nit: Make clear the IRModule can still be cross-target, right? Ie all of these hooks are responsible for doing the 'just look at defns for my target' filtering?

Reply:
Yip, eventually this should just be the unified `IRModule`, ideally it'd just be calling over these functions and they pluck the relevant functions to them. Essentially you'd end up with:

```
modules = [
  target->kind->TIRToRuntime(ir_mod, target)
  for target in targets
]
```

Right at the end, though I think that's beyond the scope of this RFC to get all the way there.